File: examples/test-game/scenes/main_menu/main_menu.c
#include "main_menu.h"

float x = 20, y = 20;
float speed = 2.0f;

static struct nk_colorf bg;

#define MAX_INPUT_LENGTH 256
#define MAX_HISTORY 32

struct console {
    char input_buffer[MAX_INPUT_LENGTH];
    char history[MAX_HISTORY][MAX_INPUT_LENGTH];
    int history_count;
    int scroll_to_bottom;
    // You can add more fields for command handling, e.g., command functions
};

struct console my_console = {0};
struct console my_console1 = {0};

EXPORT void Init(Engine_t *engine) {
    // Initialization logic for the main menu
    Engine_RunHook(engine, "HOOK_BEFORE_MAIN_MENU_INIT");
    bg = ColorToNuklearF(SKYBLUE);
    Engine_RunHook(engine, "HOOK_AFTER_MAIN_MENU_INIT");
}

EXPORT void ProcessInput(Engine_t *engine) {
    Engine_RunHook(engine, "HOOK_BEFORE_MAIN_MENU_PROCESS_INPUT");
    Engine_RunHook(engine, "HOOK_AFTER_MAIN_MENU_PROCESS_INPUT");
}

EXPORT void Update(Engine_t *engine) {
    // Update logic for the main menu
    Engine_RunHook(engine, "HOOK_BEFORE_MAIN_MENU_UPDATE");
    // printf("Main menu game update!\n");
    Engine_RunHook(engine, "HOOK_AFTER_MAIN_MENU_UPDATE");
}

EXPORT void Draw(Engine_t *engine) {
    // Drawing logic for the main menu
    Engine_RunHook(engine, "HOOK_BEFORE_MAIN_MENU_DRAW");

    struct nk_context *ctx = engine->nk_ctx;

    if (nk_begin(ctx, "Console", nk_rect(50, 50, 400, 300),
        NK_WINDOW_BORDER | NK_WINDOW_MOVABLE | NK_WINDOW_SCALABLE |
        NK_WINDOW_MINIMIZABLE | NK_WINDOW_TITLE)) {
        
        // Output history
        nk_layout_row_dynamic(ctx, 200, 1);
        if (nk_group_begin(ctx, "History", NK_WINDOW_BORDER)) {
            for (int i = 0; i < my_console.history_count; ++i) {
                nk_layout_row_dynamic(ctx, 20, 1);
                nk_label(ctx, my_console.history[i], NK_TEXT_LEFT);
            }
            if (my_console.scroll_to_bottom) {
                nk_group_set_scroll(ctx, "History", 0, my_console.history_count * 20);
                my_console.scroll_to_bottom = 0;
            }
        }
        nk_group_end(ctx);

        // Input field
        nk_layout_row_dynamic(ctx, 25, 1);
        int ret = nk_edit_string_zero_terminated(ctx, NK_EDIT_FIELD,
            my_console.input_buffer, sizeof(my_console.input_buffer), nk_filter_default);
        if (ret & NK_EDIT_ACTIVATED) {
            printf("%d\n", ret);
            printf("%d\n", NK_EDIT_ACTIVATED);
            printf("%d\n", ret & NK_EDIT_ACTIVATED);
            printf("hi\n");
        }
        if (ret & NK_EDIT_DEACTIVATED) {
            printf("%d\n", ret);
            printf("%d\n", NK_EDIT_DEACTIVATED);
            printf("%d\n", ret & NK_EDIT_DEACTIVATED);
            printf("bye\n");
        }
        if (ret & NK_EDIT_ACTIVE) {
            // printf("%s\n", ctx->text_edit.string);
            if (nk_input_is_key_pressed(&ctx->input, NK_KEY_ENTER)) {
                    strncpy(my_console.history[my_console.history_count++ % MAX_HISTORY],
                            my_console.input_buffer, MAX_INPUT_LENGTH);
                    my_console.input_buffer[0] = '\0';
                    my_console.scroll_to_bottom = 1;
            }
        }
    }
    nk_end(ctx);

    if (nk_begin(ctx, "Console1", nk_rect(50, 50, 400, 300),
        NK_WINDOW_BORDER | NK_WINDOW_MOVABLE | NK_WINDOW_SCALABLE |
        NK_WINDOW_MINIMIZABLE | NK_WINDOW_TITLE)) {
        
        // Output history
        nk_layout_row_dynamic(ctx, 200, 1);
        if (nk_group_begin(ctx, "History", NK_WINDOW_BORDER)) {
            for (int i = 0; i < my_console1.history_count; ++i) {
                nk_layout_row_dynamic(ctx, 20, 1);
                nk_label(ctx, my_console1.history[i], NK_TEXT_LEFT);
            }
            if (my_console1.scroll_to_bottom) {
                nk_group_set_scroll(ctx, "History", 0, my_console1.history_count * 20);
                my_console1.scroll_to_bottom = 0;
            }
        }
        nk_group_end(ctx);

        // Input field
        nk_layout_row_dynamic(ctx, 25, 1);
        if (nk_edit_string_zero_terminated(ctx, NK_EDIT_FIELD,
            my_console1.input_buffer, sizeof(my_console1.input_buffer), nk_filter_default) & NK_EDIT_ACTIVE) {
            // printf("%s\n", ctx->text_edit.string);
            if (nk_input_is_key_pressed(&ctx->input, NK_KEY_ENTER)) {
                    strncpy(my_console1.history[my_console1.history_count++ % MAX_HISTORY],
                            my_console1.input_buffer, MAX_INPUT_LENGTH);
                    my_console1.input_buffer[0] = '\0';
                    my_console1.scroll_to_bottom = 1;
            }
        }
    }
    nk_end(ctx);

    if (nk_begin(engine->nk_ctx, "Demo", nk_rect(50, 50, 230, 250),
                 NK_WINDOW_BORDER | NK_WINDOW_MOVABLE | NK_WINDOW_SCALABLE |
                     NK_WINDOW_MINIMIZABLE | NK_WINDOW_TITLE)) {
        enum { EASY, HARD };
        static int op = EASY;
        static int property = 20;
        nk_layout_row_static(engine->nk_ctx, 30, 80, 1);
        if (nk_button_label(engine->nk_ctx, "button"))
            TraceLog(LOG_INFO, "button pressed");

        nk_layout_row_dynamic(engine->nk_ctx, 30, 2);
        if (nk_option_label(engine->nk_ctx, "easy", op == EASY))
            op = EASY;
        if (nk_option_label(engine->nk_ctx, "hard", op == HARD))
            op = HARD;

        nk_layout_row_dynamic(engine->nk_ctx, 25, 1);
        nk_property_int(engine->nk_ctx, "Compression:", 0, &property, 100, 10,
                        1);

        nk_layout_row_dynamic(engine->nk_ctx, 20, 1);
        nk_label(engine->nk_ctx, "background:", NK_TEXT_LEFT);
        nk_layout_row_dynamic(engine->nk_ctx, 25, 1);
        if (nk_combo_begin_color(
                engine->nk_ctx, nk_rgb_cf(bg),
                nk_vec2(nk_widget_width(engine->nk_ctx), 400))) {
            nk_layout_row_dynamic(engine->nk_ctx, 120, 1);
            bg = nk_color_picker(engine->nk_ctx, bg, NK_RGBA);
            nk_layout_row_dynamic(engine->nk_ctx, 25, 1);
            bg.r = nk_propertyf(engine->nk_ctx, "#R:", 0, bg.r, 1.0f, 0.01f,
                                0.005f);
            bg.g = nk_propertyf(engine->nk_ctx, "#G:", 0, bg.g, 1.0f, 0.01f,
                                0.005f);
            bg.b = nk_propertyf(engine->nk_ctx, "#B:", 0, bg.b, 1.0f, 0.01f,
                                0.005f);
            bg.a = nk_propertyf(engine->nk_ctx, "#A:", 0, bg.a, 1.0f, 0.01f,
                                0.005f);
            nk_combo_end(engine->nk_ctx);
        }
    }
    nk_end(engine->nk_ctx);

    Engine_RunHook(engine, "HOOK_AFTER_MAIN_MENU_DRAW");
}

EXPORT void Cleanup(Engine_t *engine) {
    // Cleanup logic for the main menu
    Engine_RunHook(engine, "HOOK_BEFORE_MAIN_MENU_CLEANUP");
    Engine_RunHook(engine, "HOOK_AFTER_MAIN_MENU_CLEANUP");
}

File: examples/holojam/mods/core-mod/main_menu.lua
local cwd = function()
  local info = debug.getinfo(1, "S")
  if info and info.source then
    local path = info.source:gsub("^[@=]", "")
    return path:match("^(.*[\\/])")
  end
  return nil
end

if not WORLD then
  ECS.utils.loadNamespace(cwd() .. "components")
  local Systems = {}
  ECS.utils.loadNamespace(cwd() .. "systems", Systems)

  WORLD = ECS.world()

  local system_sort = {}
  for _, system in pairs(Systems) do
    table.insert(system_sort, system)
  end

  table.sort(system_sort, function(a, b)
    if a.priority and b.priority then
      return a.priority < b.priority
    elseif a.priority then
      return true
    else
      return false
    end
  end)

  for _, system in ipairs(system_sort) do
    WORLD:addSystem(system)
  end
end

if not CAMERA then
  CAMERA = ECS.entity(WORLD)
end
CAMERA
    :ensure("position", 0, 0)
    :ensure("velocity", 0, 0)
    :give("speed", 2.0)
    :ensure("follow")

local bg = Color.new(255, 41, 55, 255)

if not PLAYER then
  PLAYER = ECS.entity(WORLD)
end
PLAYER
    :ensure("animator", import("animators/player"):Init())
    :ensure("position", 20, 20)
    :ensure("velocity", 0, 0)
    :give("speed", 4.0)
    :give("box_collider", Vector2.new(40, 55), Vector2.new(30, 45))
    :give("controllable")
    :give("player_input", {
      up = "KEY_W",
      down = "KEY_S",
      left = "KEY_A",
      right = "KEY_D",
    })
CAMERA.follow.target = PLAYER

if not BOX then
  BOX = ECS.entity(WORLD)
end
BOX
    :ensure("position", 100, 200)
    :give("box_draw", 50, 50, Color.RED)
    :give("box_collider", Vector2.new(50, 50), Vector2.zero)

local STYLE_BLUE = import("styles/blue")
local STYLE_RED = import("styles/red")

local test_text = Engine_LoadResource("assets/test.txt", 0)
print(test_text)

MAX_HISTORY = 32
MAX_INPUT_LENGTH = 256

local debug_console = {
  window_width = 0,
  window_height = 0,
  input_buffer = "",
  io_buffer = {},
  history = {},
  history_count = 0,
  current_command = 0,
  scroll_to_bottom = false
}

RegisterFunction("HOOK_MAIN_MENU_INIT", function()
  print("(Core Mod): Main menu init!")

  local host = "localhost"
  local post = "3012"
  local ws_url = string.format("ws://%s:%s", host, post)

  local uuid = generate_uuid()
  local room = "gg"

  rtc_set_message_opened_callback(function()
    print("data channel opened")
  end)

  rtc_set_message_received_callback(function(id, message, size, ptr)
    print("data channel received"..message)
  end)

  rtc_set_message_closed_callback(function()
    print("data channel closed")
  end)

  rtc_initialize(ws_url, uuid, room)

  rtc_handle_connection()

  event_register("TEST", "main_menu_listener", function()
    print("main_menu TEST event fired")
    return false
  end)

  event_register("TEST", "main_menu_listener the second", function(type, sender, listener, event_context, context)
    print("main_menu " .. type .. " event fired by " .. sender .. " listening from " .. listener)
    return false
  end)
end)

RegisterFunction("HOOK_MAIN_MENU_PROCESS_INPUT", function()
  if IsKeyDown("KEY_BACKSLASH") then
    Engine_Mod_Scene_Switch("SETTINGS_MENU")
  end

  if EDIT_ACTIVE then
    return
  end

  if IsKeyPressed("KEY_R") then
    rtc_send_message("hi")
  end

  if IsKeyDown("KEY_F") then
    event_fire("TEST", "main_menu_sender")
  end

  if IsKeyDown("KEY_G") then
    print(bg)
  end

  if IsKeyPressed("KEY_M") then
    Engine_SetMasterVolume(0.0)
  end
  if IsKeyPressed("KEY_U") then
    Engine_SetMasterVolume(0.2)
  end
  if IsKeyPressed("KEY_I") then
    Engine_SetMasterVolume(1.0)
  end

  if IsKeyPressed("KEY_TAB") then
    DEBUG = not DEBUG
  end

  if IsKeyPressed("KEY_MINUS") then
    local group = audio_group_get(0)
    local vol = audio_group_get_volume(group)
    vol = tonumber(string.format('%.2f', vol - 0.1)) or 0
    if vol >= 0.0 then
      audio_group_set_volume(group, vol)
    end
  end
  if IsKeyPressed("KEY_EQUAL") then
    local group = audio_group_get(0)
    local vol = audio_group_get_volume(group)
    vol = tonumber(string.format('%.2f', vol + 0.1)) or 0
    if vol <= 1.0 then
      audio_group_set_volume(group, vol)
    end
  end
end)

RegisterFunction("HOOK_MAIN_MENU_UPDATE", function()
  WORLD:emit("update")
end)

RegisterFunction("HOOK_MAIN_MENU_DRAW", function()
  WORLD:emit("draw")

  if DEBUG then
    WORLD:emit("debug_draw")
  end

  -- DrawRectangle(player.x, player.y, 20, 20, bg)
  nk_style_from_table(STYLE_BLUE);
  if nk_begin("Main menu", nk_rect.new(0, 0, Engine_GetWidth(), Engine_GetHeight()), NK.WINDOW_BACKGROUND) then
    EASY = 0
    HARD = 1
    OP = OP or EASY
    PROPERTY = PROPERTY or 20
    nk_layout_row_static(30, 80, 1)
    if nk_button_label("button") then
      print("button pressed")
      rtc_send_message("message")
    end
    nk_layout_row_dynamic(30, 2);
    if nk_option_label("easy", OP == EASY) then
      OP = EASY
    end
    if nk_option_label("hard", OP == HARD) then
      OP = HARD
    end
    nk_layout_row_dynamic(25, 1)
    PROPERTY = nk_property_int("Compression:", 0, PROPERTY, 100, 10, 1)
    nk_layout_row_dynamic(20, 1)
    nk_label("background:", 3)
    nk_layout_row_dynamic(25, 1)
    if nk_combo_begin_color(bg, nk_vec2.new(nk_widget_width(), 400)) then
      nk_layout_row_dynamic(120, 1)
      bg = nk_color_picker(bg, 1);
      nk_layout_row_dynamic(25, 1)
      bg.r = nk_property_int("#R:", 0, bg.r, 255, 5, 1)
      bg.g = nk_property_int("#G:", 0, bg.g, 255, 5, 1)
      bg.b = nk_property_int("#B:", 0, bg.b, 255, 5, 1)
      bg.a = nk_property_int("#A:", 0, bg.a, 255, 5, 1)
      nk_combo_end()
    end
  end
  nk_end()

  STYLE_RED.NK_COLOR_HEADER = bg
  nk_style_from_table(STYLE_RED)
  if (nk_begin("Console", nk_rect.new(50, 50, 400, 300),
        NK.WINDOW_BORDER | NK.WINDOW_MOVABLE | NK.WINDOW_SCALABLE |
        NK.WINDOW_MINIMIZABLE | NK.WINDOW_TITLE)) then
    local line_height = 10

    if debug_console.window_height ~= nk_window_get_height() then
      local _, scroll_y = nk_group_get_scroll("History")
      nk_group_set_scroll("History", 0, scroll_y + debug_console.window_height - nk_window_get_height())
    end

    -- Output history
    nk_layout_row_dynamic(nk_window_get_height() - 80, 1)
    if (nk_group_begin("History", NK.WINDOW_BORDER)) then
      for i = 1, #debug_console.io_buffer do
        nk_layout_row_dynamic(line_height, 1)
        nk_label(debug_console.io_buffer[i], NK.TEXT_LEFT)
      end
      if (debug_console.scroll_to_bottom) then
        nk_group_set_scroll("History", 0, #debug_console.io_buffer * line_height)
        debug_console.scroll_to_bottom = false
      end
    end
    nk_group_end()

    -- Input field
    nk_layout_row_dynamic(25, 1);
    local command, state = nk_edit_string_zero_terminated(NK.EDIT_FIELD, debug_console.input_buffer, MAX_INPUT_LENGTH)
    debug_console.input_buffer = command

    if (state & NK.EDIT_ACTIVE) ~= 0 then
      if (nk_input_is_key_pressed(NK.KEY_DOWN)) then
        if debug_console.current_command + 1 <= debug_console.history_count then
          debug_console.current_command = debug_console.current_command + 1
          debug_console.input_buffer = debug_console.history[debug_console.current_command]
          nk_edit_set_cursor(debug_console.input_buffer:len())
        elseif debug_console.current_command + 1 == debug_console.history_count + 1 then
          debug_console.current_command = debug_console.current_command + 1
          debug_console.input_buffer = ""
        end
      end
      if (nk_input_is_key_pressed(NK.KEY_UP)) then
        local count = ((debug_console.current_command - 1) % MAX_HISTORY)
        if debug_console.current_command > 0 and debug_console.history[count] ~= nil then
          debug_console.current_command = debug_console.current_command - 1
          debug_console.input_buffer = debug_console.history[count]
        end
        nk_edit_set_cursor(debug_console.input_buffer:len())
      end
      if (nk_input_is_key_pressed(NK.KEY_ENTER)) then
        local argc = {}
        local regex = ("([^%s]+)"):format(" ")
        for each in command:gmatch(regex) do
          table.insert(argc, each)
        end

        table.insert(debug_console.io_buffer, "> " .. command)

        if argc[1] == "set" and #argc == 3 then
          local val = tonumber(argc[3])

          if val then
            load(argc[2] .. " = " .. argc[3])()
          else
            val = argc[3]
            local len = #val
            if val == "true" or val == "false" then
              load(argc[2] .. " = " .. val)()
            elseif val:sub(1, 1) == '"' and val:sub(len, len) == '"' then
              load(argc[2] .. " = " .. val:sub(2, len - 1))()
            end
          end
        elseif argc[1] == "get" and #argc == 2 then
          local val = _G
          regex = ("([^%s]+)"):format(".")
          for each in argc[2]:gmatch(regex) do
            val = val[each]
          end
          table.insert(debug_console.io_buffer, tostring(val))
        else
          table.insert(debug_console.io_buffer, '"' .. argc[1] .. '"' .. " command not found")
        end

        debug_console.history[(debug_console.history_count % MAX_HISTORY) + 1] = command
        debug_console.history_count = debug_console.history_count + 1
        debug_console.current_command = debug_console.history_count + 1
        debug_console.input_buffer = '\0'
        debug_console.scroll_to_bottom = true
      end
    end

    if (state & NK.EDIT_ACTIVATED) ~= 0 then
      PLAYER:remove("controllable")
      EDIT_ACTIVE = true
    end

    if (state & NK.EDIT_DEACTIVATED) ~= 0 then
      PLAYER:give("controllable")
      EDIT_ACTIVE = false
    end

    debug_console.window_width = nk_window_get_width()
    debug_console.window_height = nk_window_get_height()
  end
  nk_end()
end)

RegisterFunction("HOOK_MAIN_MENU_CLEANUP", function()
  rtc_cleanup()
  print("(Core Mod): Main menu cleanup!")
end)
